numpy/lib/arraypad.py

```diff
@@ -1193,192 +786,128 @@ def pad(array, pad_width, mode, **kwargs):
            [100, 100, 100, 100, 100, 100, 100],
            [100, 100, 100, 100, 100, 100, 100]])
     """
-    if not np.asarray(pad_width).dtype.kind == 'i':
+    array = np.asarray(array)
+    pad_width = np.asarray(pad_width)
+
+    if not pad_width.dtype.kind == 'i':
         raise TypeError('`pad_width` must be of integral type.')
 
-    narray = np.array(array)
-    pad_width = _validate_lengths(narray, pad_width)
+    # Broadcast to shape (array.ndim, 2)
+    pad_width = _as_pairs(pad_width, array.ndim, as_index=True)
 
-    allowedkwargs = {
+    if not isinstance(mode, np.compat.basestring):
+        # Old behavior: Use user-supplied function with np.apply_along_axis
+        function = mode
+        # Create a new zero padded array
+        padded, _ = _pad_simple(array, pad_width, fill_value=0)
```

Comment:
I'm not sure what is faster. Recursively calling this function with 
```
        newmat = pad(narray, pad_width=pad_width,
                     mode='constant', constant_values= 0)
```
or using `np.fill`. It probably depends a little bit on the array.

Reply:
Do you mean translating `pad(array, pad_width=..., mode="constant", constant_values=0)` to either

* `_pad_simple(array, pad_width, fill_value=constant_values)`
* vs `_pad_simple(array, pad_width)` + the iterative approach?

Yeah, that's a thought I had as well. However that is only possible if `constant_values` is a scalar otherwise we may need to pad with different values which is not possible with `_pad_simple` / `np.fill`. I think expect the fill approach to be significantly faster than the iterative one. So it might be worth catching and accelerating this (I think) really common case. Will investigate.

Reply:
hmm ok. I wouldn't worry too much about it. It was just one of the places where your code and mine had conceptually differed.